design/tlog-spilling.md.html

```diff
@@ -0,0 +1,590 @@
+*                 |       |      |          |        |       *
+*                 v       v      v         +----------+      *
+*    in-memory  +---+   +---+  +---+       |Spill Loop|      *
+*    queues     | 1 |   | 2 |  | 3 |       +----------+      *
+*    per-tag    |   |   |   |  |   |            ^            *
+*               |...|   |...|  |...|            |            *
+*                 |       |      |              |            *
+*                 v       v      v              |            *
+*                 +-------+------+--------------+            *
+*                    queues spilled on overflow              *
+*                                                            *
+**************************************************************
+
+Unfortunately, it turned out that spilling in this fashion greatly impacts TLog
+performance.  A write bandwidth saturation test was run against a cluster, with
+a modification to the transaction logs to have them act as if there was one
+storage server that was permanently failed; it never sent pop requests to allow
+the TLog to remove data from memory.  After 15min, the write bandwidth had
+reduced to 30% of its baseline.  After 30min, that became 10%.  After 60min,
+that became 5%.  (This is an intentional hyperbole due to the saturating write
+load.  See experiments at the end for more detail and data.)
+
+With the recent multi-DC/multi-region work, a failure of a remote data center
+would cause transaction logs to need to buffer all commits, as every commit is
+tagged as destined for the remote datacenter.  This would rapidly push
+transaction logs into a spilling regime, and thus write bandwidth would begin
+to rapidly degrade.  It is unacceptable for a remote datacenter failure to so
+drastically affect the primary datacenter's performance in the case of a
+failure, so a more performant way of spilling data is required.
+
+## Overview
+
+Whereas spill-by-value copied the entire mutation into the B-tree and removes
+it from the disk queue, spill-by-reference leaves the mutations in the disk
+queue and writes a pointer to it into the B-tree. Performance experiments
+revealed that the TLog's performance while spilling was dictated more by the
+number of writes done to the SQLite B-tree, than by the size of those writes.
+Thus, "spill-by-reference" being able to do a significantly better batching
+with its writes to the B-tree is more important than that it writes less data
+in aggregate.  Spill-by-reference significantly reduces the volume of data
+written to the B-tree, and the less data that we write, the more we can batch
+versions to be written together.
+
+************************************************************************
+*                              DiskQueue                               *
+*                                                                      *
+*    -------- Index on disk --------     ---- Index in memory ----     *
```

Comment:
Took me a little while to really grok this image. I think I was confused by the use of the word "Index" which made me think somehow the diskqueue contained indexing data.

I think it would be a little clear if "Index on disk" became "Entries indexed by tag in the spilled data BTree" and if "Index in memory" became "Entries indexed in memory via the by-tag queues" (although this is a little confusing because I thought the in memory queues contained the actual transactions, not pointers into the disk queue)

Also maybe a few sentences at the bottom of the diagram: "The DiskQueue is ordered by version from lowest to highest. The head of the file will contain data that was "spilled" to disk because the storage nodes it was destined for were not available at the time. These "spilled" entries are indexed by the BTree dedicated to indexing the location of spilled entries. The tail of the file contains the most recent entries and as a result will mostly contain entries that are indexed in memory via the per-tag queues."